src/IRMatch.h

```diff
@@ -2080,6 +2282,39 @@ HALIDE_ALWAYS_INLINE auto cast(halide_type_t t, A &&a) noexcept -> CastOp<declty
     return {t, pattern_arg(a)};
 }
 
+// A node for expressing type hints, when rules are ambiguously typed.
+template<typename A>
+struct TypeHint {
```

Comment:
If possible, make the type a template parameter instead of a member (either a C++ type or a halide_type_t), because each pattern consumes stack space in methods that use rewriters.